gbus/deduplicator/implementation/tx.go

```diff
@@ -0,0 +1,138 @@
+package implementation
+
+import (
+	"database/sql"
+	"time"
+
+	"emperror.dev/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/wework/grabbit/gbus"
+	"github.com/wework/grabbit/gbus/deduplicator"
+	"github.com/wework/grabbit/gbus/tx"
+)
+
+var _ deduplicator.Store = &deduper{}
+
+type deduper struct {
+	*gbus.Glogged
+	svcName    string
+	policy     gbus.DeduplicationPolicy
+	txProvider gbus.TxProvider
+	age        time.Duration
+	ticker     *time.Ticker
+	done       chan bool
+	tableName  string
+}
+
+func (d *deduper) Purge() (err error) {
+	truncateSQL := "TRUNCATE TABLE " + d.tableName
+	txp, err := d.txProvider.New()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			serr := txp.Rollback()
+			err = errors.Append(err, serr)
+		}
+		err = txp.Commit()
+	}()
+	_, err = txp.Exec(truncateSQL)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *deduper) Start() {
+	d.ticker = time.NewTicker(time.Minute)
+	d.done = make(chan bool)
+	deleteQuery := "DELETE FROM " + d.tableName + " WHERE `created_at` < ?"
+	go func() {
+		for {
+			select {
+			case <-d.done:
+				return
+			case <-d.ticker.C:
+				oldest := time.Now().Add(-1 * d.age)
+				tx, err := d.txProvider.New()
+				if err != nil {
+					d.Log().WithError(err).Error("failed to acquire a tx")
+					continue
+				}
+				result, err := tx.Exec(deleteQuery, oldest)
+				if err != nil && err != sql.ErrNoRows {
+					d.Log().WithError(err).Error("failed executing delete query")
+				}
+				n, err := result.RowsAffected()
+				if err != nil {
+					d.Log().WithError(err).Error("failed to get count of affected rows")
+				} else {
+					d.Log().WithField("table_name", d.tableName).WithField("rows_deleted", n).
+						Info("successfully cleanup duplicates table")
+				}
+			}
+		}
+	}()
+}
+
+func (d *deduper) Stop() {
+	d.Log().Info("shutting down deduplicator")
+	d.ticker.Stop()
+	close(d.done)
+}
+
+//
+func (d *deduper) StoreMessageID(tx *sql.Tx, id string) error {
```

Comment:
did you consider to store the message in another storage mechanism? like redis etc.

Reply:
I didn't want to introduce another infrastructure dependency but we can have different storages that would implement the same interface.